test/classes.coffee

```diff
@@ -1367,19 +1367,330 @@ test "nested classes with super", ->
-          # @label = super + ' from this'
+          tmp = super()
+          @label = tmp.label + ' from this'
       new ExtendedSelf
 
   eq (new Outer).label, 'outer'
   eq (new Outer.Inner).label, 'inner'
-  # eq (new Outer.ExtendedInner).label, 'inner extended'
-  # eq (new Outer.extender()).label, 'inner from this'
+  eq (new Outer.ExtendedInner).label, 'inner extended'
+  eq (Outer.extender()).label, 'outer from this'
+
+# TODO: Currently this test fails.
+test "Static methods generate 'static' keywords", ->
+  compile = """
+  class CheckStatic
+    constructor: (@drink) ->
+    @className: 'CheckStatic'
+
+  c = new CheckStatic('Machiato')
+  """
+  result = CoffeeScript.compile compile, bare: yes
+  ok (result && result.match(' static ') && true) isnt true
+
+test "Static methods in nested classes", ->
+  class Outer
+    @name: -> 'Outer'
+
+    class @Inner
+      @name: -> 'Inner'
+
+  eq Outer.name(), 'Outer'
+  eq Outer.Inner.name(), 'Inner'
+
+
+test "mixed constructors with inheritance and ES6 super", ->  
+  identity = (f) -> f
+
+  class TopClass
+    constructor: (arg) ->
+      @prop = 'top-' + arg
+
+  ```
+  class SuperClass extends TopClass {
+    constructor (arg) {
+      identity(super('super-' + arg));
+    }
+  }
+  ```
+  class SubClass extends SuperClass
+    constructor: ->
+      identity super 'sub'
+
+  ok (new SubClass).prop is 'top-super-sub'
+
+test "ES6 static class methods can be overriden", ->
+  class A
+    @name: -> 'A'
+
+  class B extends A
+    @name: -> 'B'
+
+  eq A.name(), 'A'
+  eq B.name(), 'B'
+
+# If creating static by direct assignment rather than ES6 static keyword
+test "ES6 Static methods should set `this` to undefined // ES6 ", ->
+  class A
+    @test: -> 
+      eq this, undefined
+
+# Ensure that our object prototypes work with ES6
+test "ES6 prototypes can be overriden", ->
+  class A
+    className: 'classA'
+
+  ```
+  class B {
+    test () {return "B";};
+  }
+  ```
+  b = new B
+  a = new A
+  eq a.className, 'classA'
+  eq b.test(), 'B'
+  Object.setPrototypeOf(b, a)
+  eq b.className, 'classA'
+  # This shouldn't throw, 
+  # as we only change inheritance not object construction
+  # This may be an issue with ES, rather than CS construction?
+  #eq b.test(), 'B'
+
+  class D extends B  
+  B::test = () -> 'D'
+  eq (new D).test(), 'D'
+
+
+test "ES6 conformance to extending non-classes", ->
+  A = (@title) ->
+    'Title: ' + @
+
+  class B extends A
+  b = new B('caffeinated') 
+  eq b.title, 'caffeinated'
+
+  # Check inheritance chain 
+  A::getTitle = () -> @title
+  eq b.getTitle(), 'caffeinated'
+
+  throwsC = """
+  C = {title: 'invalid'}
+  class D extends {}
+  """
+  # TODO: This should catch on compile and message should be "class can only extend classes and functions."
+  throws -> CoffeeScript.run throwsC, bare: yes
+
+# TODO: Either restrict valid statements in class bodies OR insert 'use strict'?
```

Comment:
On node side that's fine, but on the client generated code it will lead to subtle errors if we have mixed classes. 
This is what led to the test:  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#Strict_mode